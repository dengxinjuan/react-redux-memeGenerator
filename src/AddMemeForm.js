import React from "react";
import { useState } from "react";
import { v4 as uuidv4 } from "uuid"; // get uuid

const AddMemeForm = ({ add }) => {
  /**send value to parent and clear form **/
  const INITIAL_STATE = { topText: "", bottomText: "", url: "" };
  const [form, setForm] = useState(INITIAL_STATE);

  /*handle add function, which herited from parent App*/
  function handleSubmit(e) {
    e.preventDefault();
    add({ ...form, id: uuidv4() });
    setForm(INITIAL_STATE);
  }
  /*handle change when capturing the value*/
  function handleChange(e) {
    const { name, value } = e.target;
    setForm((f) => ({ ...f, [name]: value }));
  }

  return (
    <div>
      <h2>Make a New Meme</h2>
      <form onSubmit={handleSubmit}>
        <label htmlFor="url">Image URL</label>
        <input
          type="text"
          name="url"
          id="form_url"
          onChange={handleChange}
          value={form.url}
        />
        <label htmlFor="topText">Top Text</label>
        <input
          type="text"
          name="topText"
          id="form_topText"
          onChange={handleChange}
          value={form.topText}
        />
        <label htmlFor="bottomText">Bottom Text</label>
        <input
          type="text"
          name="bottomText"
          id="form_bottomText"
          onChange={handleChange}
          value={form.bottomText}
        />
        <button type="submit" id="form_submit">
          Generate Meme!
        </button>
      </form>
    </div>
  );
};

export default AddMemeForm;
