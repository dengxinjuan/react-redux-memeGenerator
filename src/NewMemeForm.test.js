import React from "react";
import { render, fireEvent } from "@testing-library/react";
import AddMemeForm from "./AddMemeForm";

let memeForm;

describe("<AddMemeForm /> basics", () => {
  beforeEach(() => {
    memeForm = render(<AddMemeForm />);
  });

  it("renders", () => {
    expect(memeForm).toBeTruthy();
  });
  it("matches snapshot", () => {
    expect(memeForm.asFragment()).toMatchSnapshot();
  });
});

it("updates state with topText input", () => {
  // mount form and mock addMeme
  const mockFn = jest.fn();
  const { getByText } = render(<AddMemeForm add={mockFn} />);

  fireEvent.click(getByText("Generate Meme!"));

  expect(mockFn).toHaveBeenCalled();
});
