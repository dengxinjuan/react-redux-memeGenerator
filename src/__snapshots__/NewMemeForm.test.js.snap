// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMemeForm /> basics matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Make a New Meme
    </h2>
    <form>
      <label
        for="url"
      >
        Image URL
      </label>
      <input
        id="form_url"
        name="url"
        type="text"
        value=""
      />
      <label
        for="topText"
      >
        Top Text
      </label>
      <input
        id="form_topText"
        name="topText"
        type="text"
        value=""
      />
      <label
        for="bottomText"
      >
        Bottom Text
      </label>
      <input
        id="form_bottomText"
        name="bottomText"
        type="text"
        value=""
      />
      <button
        id="form_submit"
        type="submit"
      >
        Generate Meme!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
